Import a file to a disk

jmpesp opened the pull request:
Use the bulk-write import functionality to upload a file to a disk,
optionally making a snapshot and an image out of that snapshot. An
example invocation:

    oxide disk import \
        --path crucible-tester-sparse.img \
        --threads 8 \
        --disk-name crucible-tester-disk \
        --project myproj \
        --description "testing import!" \
        --snapshot-name crucible-tester-snapshot \
        --image-name crucible-tester \
        --image-description "uploading of crucible-tester-sparse.img" \
        --image-os debian \
        --image-version 11

This will:

- create a disk named `crucible-tester-disk` in state "import-ready"
- set that disk to "importing-via-bulk-write"
- spawn 8 upload tasks and bulk write file chunks to the disk in
  parallel
- wait for it to be done
- stop the bulk write process
- finalize the disk, creating the snapshot `crucible-tester-snapshot`
- create the image `crucible-tester` out of that snapshot

jmpesp committed:
Import a file to a disk

Use the bulk-write import functionality to upload a file to a disk,
optionally making a snapshot and an image out of that snapshot. An
example invocation:

    oxide disk import \
        --path crucible-tester-sparse.img \
        --threads 8 \
        --disk-name crucible-tester-disk \
        --project myproj \
        --description "testing import!" \
        --snapshot-name crucible-tester-snapshot \
        --image-name crucible-tester \
        --image-description "uploading of crucible-tester-sparse.img" \
        --image-os debian \
        --image-version 11

This will:

- create a disk named `crucible-tester-disk` in state "import-ready"
- set that disk to "importing-via-bulk-write"
- spawn 8 upload tasks and bulk write file chunks to the disk in
  parallel
- wait for it to be done
- stop the bulk write process
- finalize the disk, creating the snapshot `crucible-tester-snapshot`
- create the image `crucible-tester` out of that snapshot

jmpesp committed:
regenerate cli.json

jmpesp committed:
fmt

jmpesp committed:
Merge remote-tracking branch 'origin/main' into image_import

jmpesp committed:
Merge remote-tracking branch 'origin/main' into image_import

jmpesp committed:
rework the finalize for update - snapshot name is in body now

jmpesp committed:
make project images instead of global images

jmpesp committed:
Merge remote-tracking branch 'origin/main' into image_import

jmpesp committed:
actually make a project image!

jmpesp commented:
Closes https://github.com/oxidecomputer/oxide.rs/issues/82

david-crespo commented on cli/src/cmd_disk.rs:
```diff
@@ -4,27 +4,457 @@
+}
+
+// Upload to Nexus in 512k byte chunks
+const CHUNK_SIZE: u64 = 512 * 1024;
```
This seems really small!

david-crespo replied:
Ah. It comes from Crucible: https://github.com/oxidecomputer/crucible/blob/0b34a53b57038a7804ac69ab90400539bd1961e8/pantry/src/pantry.rs#L33

ahl commented on cli/src/cmd_disk.rs:
```diff
@@ -4,27 +4,457 @@
+            let disk_name = self.disk_name.clone();
+            let project = self.project.clone();
+
+            handles.push(tokio::spawn(async move {
```
I guess a task is the same as a thread to you...

jmpesp committed:
Merge branch 'main' into image_import

jmpesp committed:
start on integration tests

jmpesp committed:
test, not text

jmpesp committed:
handle one way Option is generated

jmpesp committed:
some unwind tests

jmpesp commented:
@ahl ready for review. I've done what tests I could given that httpmock can't return different responses for the same request.

ahl commented on cli/src/cmd_disk.rs:
```diff
@@ -4,27 +4,518 @@
+    /// How many upload threads to run. Disk size must be divisible by this
+    /// number.
+    #[clap(long, default_value = "1")]
+    threads: usize,
```
This seems challenging for users; can we eliminate this constraint?

Can we rename this `concurrency`? Why would someone change this from the default? Presumably to make the data transfer faster. How would I know if the right value were 2 or 20 or 200? Is this particularly CPU intensive? That seems unlikely. Am I really trying to influence bandwidth consuption?

jmpesp replied:
I basically always set this to 8. Each of the upload tasks goes at about 5 MiB/s, so the aggregate is 40 MiB/s. Above 8 tasks, the aggregate doesn't go up. Maybe it'd be better to have one progress bar for all upload threads, set this to 8 by default, and rename it `upload-concurrency`?

I'll stare at the code a bit to see if I can remove the divisible constraint.

jmpesp replied:
3433f6675cc1861d11be9716f8b81339d8ce7bc5 does away with the `--threads` argument, hard coding it to 8. it also loosens the constraint for the user - their file size must be evenly divided by the block size. The pantry will reject writes where this isn't true anyway.

ahl commented on cli/src/cmd_disk.rs:
```diff
@@ -4,27 +4,518 @@
+    /// The size of the disk to create. If unspecified, the size of the file
+    /// will be used, rounded up to the nearest GB.
+    #[clap(long)]
+    disk_size: Option<u64>,
```
I **think** if you use `oxide_api::types::ByteCount` you'll get suffix parsing "for free"

jmpesp replied:
Done in d8837c8

ahl commented on cli/src/cmd_disk.rs:
```diff
@@ -4,27 +4,518 @@
+
+    /// The desired size of the disk's blocks.
+    #[clap(long)]
+    disk_block_size: Option<i64>,
```
I see that we default to 512 below (is that what we want?). We might want to use `oxide_api::types::BlockSize`. I'm not exactly sure of the best way to do this, but it's possible to have clap show us only the allowable values for this property in the help output. I think this may involve either adding your own enum or adding a new impl to `clap_feature.rs` for `BlockSize`

jmpesp replied:
I've done this in d8837c8, but it looks like the error messages could be better:

```
$ ./cli.sh disk import --disk-size 123a
error: one of the values isn't valid for an argument
$ ./cli.sh disk import --disk-block-size 123
error: one of the values isn't valid for an argument
```

ahl commented on cli/src/cmd_disk.rs:
```diff
@@ -4,27 +4,518 @@
+    #[clap(long)]
+    image_description: Option<String>,
+
+    /// The OS of this image (eg. Debian)
```
```suggestion
    /// The OS of this image (e.g. Debian)
```

jmpesp replied:
done in 3d5bdd8e6e968bfee866985c38ebae5044521e5f

ahl commented on cli/src/cmd_disk.rs:
```diff
@@ -4,27 +4,518 @@
+    #[clap(long)]
+    image_os: Option<String>,
+
+    /// The version of this image (eg. 11, focal, a9e77e3a, 2023-04-06T14:23:34Z)
```
```suggestion
    /// The version of this image (e.g. 11, focal, a9e77e3a, 2023-04-06T14:23:34Z)
```

jmpesp replied:
done in 3d5bdd8e6e968bfee866985c38ebae5044521e5f

ahl commented on cli/src/cmd_disk.rs:
```diff
@@ -4,27 +4,518 @@
+use std::path::Path;
+use tokio::sync::mpsc;
 
+/// Import a file to a disk
```
```suggestion
/// Create a disk from a file upload
```

maybe?

jmpesp replied:
changed in 3d5bdd8e6e968bfee866985c38ebae5044521e5f - let me know what you think

ahl commented on cli/src/cmd_disk.rs:
```diff
@@ -4,27 +4,518 @@
+/// Import a file to a disk
+///
+/// Upload the contents of a file to a disk, optionally creating a snapshot
+/// and/or an image.
```
```suggestion
/// Upload the contents of a file to create a new disk. Use the --snapshot
/// or --image options to also create a snapshot or image.
```

again: maybe?

jmpesp replied:
changed in 3d5bdd8e6e968bfee866985c38ebae5044521e5f - let me know what you think

ahl commented on cli/src/cmd_disk.rs:
```diff
@@ -4,27 +4,518 @@
+
+    /// If supplied, create a snapshot with this name.
+    #[clap(long)]
+    snapshot_name: Option<Name>,
```
```suggestion

    /// If supplied, create a snapshot with the given name.
    #[clap(long)]
    snapshot: Option<Name>,
```

jmpesp replied:
changed in 3d5bdd8e6e968bfee866985c38ebae5044521e5f

ahl commented on cli/src/cmd_disk.rs:
```diff
@@ -4,27 +4,518 @@
+    /// If supplied, create an image from the snapshot of this disk. Requires
+    /// snapshot name to be supplied.
+    #[clap(long)]
+    image_name: Option<Name>,
```
```suggestion
    /// If supplied, create an image with the given name. Requires the creation of a snapshot.
    #[clap(long, requires = "snapshot")]
    image: Option<Name>,
```

jmpesp replied:
changed in 3d5bdd8e6e968bfee866985c38ebae5044521e5f

ahl commented on cli/src/cmd_disk.rs:
```diff
@@ -4,27 +4,518 @@
+    image_name: Option<Name>,
+
+    /// The description for the image created out of the snapshot of this disk.
+    #[clap(long)]
```
```suggestion
    #[clap(long, requires = "snapshot", requires = "image")]
```

ahl replied:
It seems like image/image-name requires image-description, image-os, and image-version as well, yes? You could make a clap group to encode this.

ahl replied:
still this? ^

jmpesp replied:
see "I wasn't smart enough to get groups to work" @ 47ef018 - I used explicitly provided `requires_all`

ahl commented on cli/src/cmd_disk.rs:
```diff
@@ -4,27 +4,518 @@
+    image_version: Option<String>,
+}
+
+/// Return a correct disk size for the path and size arguments
```
What does "correct" mean in this context?

jmpesp replied:
changed in 3d5bdd8e6e968bfee866985c38ebae5044521e5f

ahl commented on cli/src/cmd_disk.rs:
```diff
@@ -4,27 +4,518 @@
+    };
+
+    // Nexus' disk size minimum is 1 GB
+    let disk_size = std::cmp::max(ONE_GB, disk_size);
```
is this needed given the code below that rounds up?

jmpesp replied:
no, changed in ffab026

ahl commented on cli/src/cmd_disk.rs:
```diff
@@ -4,27 +4,518 @@
+            return Err(e.into());
+        }
+
+        // Create one tokio task for each thread that will upload file chunks
```
would this be simpler with rayon... or more complicated?

ahl commented on cli/src/cmd_disk.rs:
```diff
@@ -4,27 +4,518 @@
+
+            // Attempt to unwind the disk, although it will probably fail - the
+            // first step is to stop the bulk write process!
+            self.unwind_disk_bulk_write_stop(&client).await?;
```
might we want to retry for a bit? Perhaps depending on the status code?

ahl commented on cli/src/cmd_disk.rs:
```diff
@@ -4,27 +4,518 @@
+            .project(&self.project)
+            .disk(self.disk_name.clone())
+            .body(FinalizeDisk {
+                snapshot_name: self.snapshot_name.as_ref().map(|x| x.clone()),
```
```suggestion
                snapshot_name: self.snapshot_name.clone(),
```
?

jmpesp replied:
cool! changed in 1670a86

ahl commented on cli/src/cmd_disk.rs:
```diff
@@ -4,27 +4,518 @@
+                .snapshot(NameOrId::Name(snapshot_name.clone()))
+                .send()
+                .await?
+                .into_inner();
```
for my education: are the explicit type and `into_inner()` call necessary here?

ahl commented on cli/src/main.rs:
```diff
@@ -137,7 +137,8 @@ async fn main() {
 
         // TODO this could be improved
```
```suggestion
```

jmpesp replied:
turns out no! ff59b45

ahl commented on cli/tests/test_disk_import.rs:
```diff
@@ -0,0 +1,550 @@
+        .arg("tests/data/testpost.iso")
+        .arg("--disk-name")
+        .arg("test-import")
+        .assert()
+        .success();
```
do you want to look at the command output? No is a fine answer ;-)

jmpesp replied:
not in this case :)

ahl commented on cli/tests/test_disk_import.rs:
```diff
@@ -0,0 +1,550 @@
+        .arg("--image-os")
+        .arg("dos")
+        .arg("--image-version")
+        .arg("6.1")
```
makes sense.

ahl commented on cli/tests/test_disk_import.rs:
```diff
@@ -0,0 +1,550 @@
+            &oxide_api::types::Error {
+                error_code: None,
+                message: "disk not found".into(),
+                request_id: Uuid::new_v4().to_string(),
```
If you want this to be predictable you **could** do `Uuid::mock_value(&mut src)`

jmpesp replied:
I like predictable, changed in 8d43e6d

ahl commented on test-common/src/json_mock.rs:
```diff
@@ -176,6 +176,24 @@ fn mock_schema_object(
             extensions: _,
         } if instance_type.as_ref() == &InstanceType::String => mock_string(format, string, src),
 
+        // UUID
```
```suggestion
        // One way to express an option
```

jmpesp replied:
done in f393d4e

ahl commented on test-common/src/json_mock.rs:
```diff
@@ -176,6 +176,24 @@ fn mock_schema_object(
+            object: _,
+            reference: None,
+            extensions: _,
+        } if matches!(&vec[..], &[InstanceType::String, InstanceType::Null]) => {
```
not sure this is strictly necessary, but for robustness:

```suggestion
        } if matches!(&vec[..], &[InstanceType::String, InstanceType::Null] | &[InstanceType::Null, InstanceType::String]) => {
```

Not required here, but I think I'll change this in the future to do something like we do in typify:

https://github.com/oxidecomputer/typify/blob/main/typify-impl/src/convert.rs#L56

jmpesp replied:
done in f393d4e

ahl commented on test-common/src/json_mock.rs:
```diff
@@ -233,6 +252,22 @@ fn mock_schema_object(
             mock_schema(schema, definitions, src)
         }
 
+        // one way that Option is generated
```
```suggestion
        // The null schema--we see this sometimes in an `anyOf` resulting from an `Option`
```

jmpesp replied:
done in f393d4e

ahl reviewed (commented):
looks good; some notes regarding args and arg parsing

jmpesp committed:
use ByteCount and BlockSize in disk import arguments

add BlockSize clap_feature.rs

jmpesp committed:
reword

jmpesp committed:
cmp::max not needed for get_disk_size

jmpesp committed:
"as_ref().map(|x| x.clone())" => "clone()"

jmpesp committed:
turns out we don't need into_inner!

jmpesp committed:
update docs

jmpesp committed:
use Uuid::mock_value

jmpesp committed:
should be speaking the truth in json_mock.rs

jmpesp committed:
fmt

jmpesp committed:
support sparse images: do not POST blocks that are all 0

jmpesp committed:
do away with --threads, always use 8

use one progress bar

enforce the constraint that file size has to be divisible by selected
block size

send the chunk size into the upload task channel, and accurately update
the progress bar

only base64 encode how many bytes we read

jmpesp committed:
beef up disk import tests with tempfiles

remove cli/tests/data/testpost.iso

ahl commented on sdk/src/clap_feature.rs:
```diff
@@ -58,3 +58,37 @@ impl clap::builder::TypedValueParser for ByteCountParser {
+        }
+
+        parse(cmd, arg, value).ok_or_else(|| clap::Error::new(clap::error::ErrorKind::InvalidValue))
+    }
```
why was this needed?

jmpesp replied:
Otherwise, I get
```
error[E0599]: the method `value_parser` exists for reference `&&&&&&_AutoValueParser<BlockSize>`, but its trait bounds were not satisfied
    --> cli/src/cmd_disk.rs:63:5
     |
63   |     /// The desired size of the disk's blocks.
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ method cannot be called on `&&&&&&_AutoValueParser<BlockSize>` due to unsatisfied trait bounds
     |
    ::: /home/jwm/.cargo/registry/src/github.com-1ecc6299db9ec823/clap_builder-4.2.4/src/builder/value_parser.rs:2221:1
     |
2221 | pub struct _AutoValueParser<T>(std::marker::PhantomData<T>);
     | ------------------------------ doesn't satisfy `_: _ValueParserViaParse`
     |
    ::: /home/jwm/src/oxidecomputer/oxide.rs/sdk/src/generated_sdk.rs:127:5
     |
127  |     pub struct BlockSize(i64);
     |     --------------------
     |     |
     |     doesn't satisfy `BlockSize: From<&'s std::ffi::OsStr>`
     |     doesn't satisfy `BlockSize: From<&'s str>`
     |     doesn't satisfy `BlockSize: From<OsString>`
     |     doesn't satisfy `BlockSize: From<std::string::String>`
     |     doesn't satisfy `BlockSize: FromStr`
     |     doesn't satisfy `BlockSize: ValueEnum`
     |     doesn't satisfy `BlockSize: ValueParserFactory`
     |
     = note: the following trait bounds were not satisfied:
             `BlockSize: ValueEnum`
             which is required by `&&&&&_AutoValueParser<BlockSize>: clap::builder::via_prelude::_ValueParserViaValueEnum`
             `BlockSize: ValueParserFactory`
             which is required by `&&&&&&_AutoValueParser<BlockSize>: clap::builder::via_prelude::_ValueParserViaFactory`
             `BlockSize: From<OsString>`
             which is required by `&&&&_AutoValueParser<BlockSize>: clap::builder::via_prelude::_ValueParserViaFromOsString`
             `BlockSize: From<&'s std::ffi::OsStr>`
             which is required by `&&&_AutoValueParser<BlockSize>: clap::builder::via_prelude::_ValueParserViaFromOsStr`
             `BlockSize: From<std::string::String>`
             which is required by `&&_AutoValueParser<BlockSize>: clap::builder::via_prelude::_ValueParserViaFromString`
             `BlockSize: From<&'s str>`
             which is required by `&_AutoValueParser<BlockSize>: clap::builder::via_prelude::_ValueParserViaFromStr`
             `BlockSize: FromStr`
             which is required by `_AutoValueParser<BlockSize>: clap::builder::via_prelude::_ValueParserViaParse`
     = note: this error originates in the macro `clap::value_parser` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0599`.
error: could not compile `oxide` due to 2 previous errors
```

ahl replied:
ok; you might want to just do the validation inline: "512", "2048", "2k" ... (not sure why we support 2k, but I'm sure there are reasons...)

jmpesp replied:
> do the validation inline

I'm not 100% sure what you mean by this but my assumption is

```patch
diff --git a/sdk/src/clap_feature.rs b/sdk/src/clap_feature.rs
index 04dfc65..9348fa7 100644
--- a/sdk/src/clap_feature.rs
+++ b/sdk/src/clap_feature.rs
@@ -84,11 +84,16 @@ impl clap::builder::TypedValueParser for BlockSizeParser {
         ) -> Option<types::BlockSize> {
             let value: Option<i64> = value.to_str()?.parse().ok();
             match value {
-                Some(v) => v.try_into().ok(),
+                Some(512) => 512.try_into().ok(),
+                Some(2048) => 2048.try_into().ok(),
+                Some(4096) => 4096.try_into().ok(),
+                Some(_) => None,
                 None => None,
             }
         }
```

which I've committed (along with a better error message) in 6847f20

ahl commented on cli/src/cmd_disk.rs:
```diff
@@ -4,27 +4,550 @@
+            );
+        }
+
+        // Use 8 upload tasks - this evenly divides all block sizes, and we know
+        // that the file size is evenly divided by the selected block size due
+        // to the above check.
+        const UPLOAD_TASKS: usize = 8;
```
love it

jmpesp replied:
:+1:

ahl reviewed (commented):
looks good to me

jmpesp committed:
correct progress bar accounting

jmpesp committed:
clippy --all-targets!

jmpesp committed:
shorter argument names

jmpesp committed:
I wasn't smart enough to get groups to work

jmpesp committed:
validate block size in BlockSizeParser

jmpesp committed:
Merge remote-tracking branch 'origin/main' into image_import

jmpesp commented:
I'm able to reproduce this test failure on my workstation, but it's intermittent!

```
---- test_disk_import_bulk_write_import_fail stdout ----
thread 'test_disk_import_bulk_write_import_fail' panicked at 'assertion failed: `(left == right)`
  left: `"^/v1/disks/.*/bulk-write$"`,
 right: `"/v1/disks/test-import"`: 1 of 2 expected requests matched the mock specification, .
Here is a comparison with the most similar non-matching request (request number 1): 

1 : The path does not match
------------------------------------------------------------------------------------------
Expected:	[matches regex]		^/v1/disks/.*/bulk-write$
Actual:		                    	/v1/disks/test-import

2 : The method does not match
------------------------------------------------------------------------------------------
Expected:	[equals]		POST
Actual:		             	GET

jwm@duster ~/src/oxidecomputer/oxide.rs [image_import] [] $ cargo test --all-features
    Finished test [unoptimized + debuginfo] target(s) in 0.40s
     Running unittests src/main.rs (target/debug/deps/oxide-befbc3f1a577763c)

running 6 tests
test cmd_auth::test::test_parse_host ... ok
test tests::test_autoref ... ok
test cmd_api::test::test_add_query_string ... ok
test cmd_disk::test_get_disk_size ... ok
test cmd_version::version_success ... ok
test cmd_auth::test_cmd_auth_status ... ok

test result: ok. 6 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.54s

     Running tests/test_api.rs (target/debug/deps/test_api-24c40ef8fd905d65)

running 3 tests
test test_pagination_midway_failure ... ok
test test_pagination_success ... ok
test test_simple_api_call ... ok

test result: ok. 3 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.09s

     Running tests/test_disk_import.rs (target/debug/deps/test_disk_import-0bebbfea6cf5888d)

running 10 tests
test test_disk_import_bad_file_size ... ok
test test_disk_import_bad_block_size ... ok
test test_disk_import_disk_exists_already ... ok
test test_disk_import_bulk_import_start_fail ... ok
test test_disk_import_image_exists_already ... ok
test test_disk_import_required_parameters ... ok
test test_disk_import_snapshot_exists_already ... ok
test test_disk_import_sparse ... ok
test test_disk_import_bulk_write_import_fail ... ok
test test_disk_import ... ok

...
```

ahl commented on cli/tests/test_disk_import.rs:
```diff
@@ -0,0 +1,772 @@
+// A disk import where the requested block size is invalid
+#[test]
+fn test_disk_import_bad_block_size() {
+    let server = MockServer::start();
```
unneeded?

ahl commented on cli/tests/test_disk_import.rs:
```diff
@@ -0,0 +1,772 @@
+// A disk import where the file size doesn't divide the block size
+#[test]
+fn test_disk_import_bad_file_size() {
+    let server = MockServer::start();
```
unneeded?

ahl commented on cli/tests/test_disk_import.rs:
```diff
@@ -0,0 +1,772 @@
+// Test for required parameters being supplied
+#[test]
+fn test_disk_import_required_parameters() {
+    let server = MockServer::start();
```
unneeded?

jmpesp committed:
trying things to isolate why tests are failing

- each test now uses a different rand::rngs::SmallRng::seed_from_u64
- each test now uses a different bearer token
- each test now uses different disk, snapshot, and image names
- each test now prints the server url it uses
- tests that don't need MockServer now don't create one

jmpesp committed:
let all upload tasks / threads complete

previously, the cli would bail out too early

**this** was the cause of the test instability: the cli would tear itself down and only 1 request would be sent

jmpesp committed:
rip out different Authorization headers

ahl reviewed (approved):
I approve per your recent threats.
